### PR TITLE
Monitoring: add expire option for comments

### LINKED ISSUE #3447: Expire Option for Comments
## Expected Behavior

Comments should disappear when they become irrelevant and stay visible just in the event history. 

## Current Behavior

Comments stay visible at the commented Event forever if you don't delete them manually. 

## Possible Solution

Implement an "expire date" option for comments like the expire option for acknowledgements. 

## Context

We are trying to avoid having old and confusing information at some Icinga events. 
Additionally we are trying to avoid having to delete all those old comments manually. 

## Your Environment
Icinga Web2 Version 2.5.0

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #3447

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] Icinga/icinga2#8035

**Comment by Al2Klimov** [comment]
Bildschirmfoto 2020-05-27 um 13 31 31

File: modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php [hunk 1]
```diff
@@ -63,6 +65,52 @@ public function createElements(array $formData = array())
...
+        );
+
+        if (isset($formData['expire']) ? $formData['expire'] : $commentExpire) {
+            $expireTime = new DateTime();
+            $expireTime->add(new DateInterval($config->get('settings', 'comment_expire_time', 'PT1H')));
+
+            $this->addElement(
+                'dateTimePicker',
+                'expire_time',
+                array(
```
Comment (nilmerg) [inline modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php:92]: Please use `[` and `]` instead.

File: modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php [hunk 2]
```diff
@@ -63,6 +65,52 @@ public function createElements(array $formData = array())
...
+                    'value'         => $expireTime,
+                    'description'   => $this->translate(
+                        'Enter the expire date and time for this comment here. Icinga will delete the'
+                        . ' comment after this time expired.'
+                    )
+                )
+            );
+
+            $this->addDisplayGroup(
+                array('expire', 'expire_time'),
```
Comment (nilmerg) [inline modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php:103]: Please use `[` and `]` instead.

File: modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php [hunk 3]
```diff
@@ -63,6 +65,52 @@ public function createElements(array $formData = array())
...
+                'expire-expire_time',
+                array(
```
Comment (nilmerg) [inline modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php:105]: Please use `[` and `]` instead.

File: modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php [hunk 4]
```diff
@@ -63,6 +65,52 @@ public function createElements(array $formData = array())
...
+                    'decorators' => array(
+                        'FormElements',
+                        array('HtmlTag', array('tag' => 'div'))
```
Comment (nilmerg) [inline modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php:108]: Please use `[` and `]` instead.

File: modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php [hunk 5]
```diff
@@ -81,6 +129,9 @@ public function onSuccess()
             if (($persistent = $this->getElement('persistent')) !== null) {
                 $comment->setPersistent($persistent->isChecked());
             }
+            if ($this->getElement('expire')->isChecked()) {
+                $comment->setExpireTime($this->getElement('expire_time')->getValue()->getTimestamp());
+            }
```
Comment (nilmerg) [inline modules/monitoring/application/forms/Command/Object/AddCommentCommandForm.php:134]: Please add an empty line before and after this block. We require an empty line after closing braces.

File: modules/monitoring/application/views/scripts/show/components/comments.phtml [hunk 6]
```diff
@@ -50,6 +50,13 @@ if (empty($object->comments) && ! $addLink) {
                 <span class="comment-time">
                     <?= $this->translate('commented') ?>
                     <?= $this->timeAgo($comment->timestamp) ?>
+                <?php if ($comment->expiration): ?>
+                    <span aria-hidden="true">&#448;</span>
```
Comment (nilmerg) [inline modules/monitoring/application/views/scripts/show/components/comments.phtml:54]: Why are you using the encoded version of `|` here? It's useless and already cost me too much time looking up what this code represents. Use `|` instead.

File: modules/monitoring/library/Monitoring/Command/Object/AddCommentCommand.php [hunk 7]
```diff
@@ -46,4 +53,27 @@ public function getPersistent()
...
+     * Set the time when the acknowledgement should expire
+     *
+     * @param   int $expireTime
+     *
+     * @return  $this
+     */
+    public function setExpireTime($expireTime)
+    {
+        $this->expireTime = (int) $expireTime;
+        return $this;
```
Comment (nilmerg) [inline modules/monitoring/library/Monitoring/Command/Object/AddCommentCommand.php:67]: Please add an empty line before the return statement. We require empty lines before return statements.

File: modules/monitoring/library/Monitoring/Command/Renderer/IcingaApiCommandRenderer.php [hunk 8]
```diff
@@ -99,6 +99,9 @@ public function renderAddComment(AddCommentCommand $command)
             'author'    => $command->getAuthor(),
             'comment'   => $command->getComment()
         );
+        if ($command->getExpireTime() !== null) {
+            $data['expiry'] = $command->getExpireTime();
+        }
```
Comment (nilmerg) [inline modules/monitoring/library/Monitoring/Command/Renderer/IcingaApiCommandRenderer.php:104]: Please add an empty line after this block. We require an empty line after closing braces.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Please also check Icinga's version and only show the option to define an expire time if it's `>=2.13`

**Comment by Al2Klimov** [comment]
Do you fear Icinga 2 API request failures otherwise?

**Comment by nilmerg** [comment]
I fear users checking that box but not having Icinga 2.13 yet, and complaining then it doesn't work.

File: modules/monitoring/application/views/scripts/show/components/comments.phtml (hunk 6 above)
Comment (Al2Klimov) [reply modules/monitoring/application/views/scripts/show/components/comments.phtml:54]: C&P from modules/monitoring/application/views/scripts/show/components/acknowledgement.phtml

**Comment by Al2Klimov** [comment]
Interesting... why did you put this on v2.9? Neither the Icinga 2 counterpart has been merged, nor v2.13's release date is known, yet.

**Comment by nilmerg** [comment]
Once you check for that version in code, it doesn't even show up until 2.13 is released, but if Icinga Web 2.9 is released earlier, it just shows magically up once the user upgrades Icinga 2. :mage:

**Review by nilmerg — CHANGES_REQUESTED** [review]
Code looks good now. Though please also include the new default configuration option in [`modules/monitoring/doc/03-Configuration.md`](https://github.com/Icinga/icingaweb2/blob/master/modules/monitoring/doc/03-Configuration.md#default-settings-).

**Review by nilmerg — CHANGES_REQUESTED** [review]
You've introduced *two* new default options if I'm not mistaken.

**Comment by Al2Klimov** [comment]
Anyway:

> Neither the Icinga 2 counterpart has been merged

**Review by nilmerg — APPROVED** [review]

**Pull request merged**